src/rust/client-executor/src/lib.rs

```diff
@@ -0,0 +1,421 @@
+            Some(duration) => {
+                tracing::debug!("Retrying in {}ms", duration.as_millis());
+                let deadline = Instant::now() + duration;
+                let future = sleep_until(deadline);
+                self.as_mut()
+                    .project()
+                    .state
+                    .set(ExecuteState::Sleeping(future));
+                Ok(self.poll(cx))
+            }
+        }
+    }
+}
+
+impl<'a, I, A> Future for Execution<'a, I, A>
+where
+    I: Iterator<Item = Duration>,
+    A: Action,
+{
+    type Output = ActionResult<A>;
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
+        match self.as_mut().project().state.poll(cx) {
+            ExecuteFuturePoll::Running(poll_result) => match poll_result {
+                Poll::Ready(Ok(ok)) => Poll::Ready(Ok(ok)),
+                Poll::Pending => Poll::Pending,
+                Poll::Ready(Err(err)) => match self.retry(err, cx) {
+                    Ok(poll) => poll,
+                    Err(err) => Poll::Ready(Err(err)),
+                },
+            },
+            ExecuteFuturePoll::Sleeping(poll_result) => match poll_result {
+                Poll::Pending => Poll::Pending,
+                Poll::Ready(_) => self.attempt(cx),
+            },
+        }
+    }
+}
+
+/// Based on https://resilience4j.readme.io/docs/circuitbreaker
+pub struct ExecutorConfig {
+    builder: RecloserBuilder,
+    timeout: Duration,
+}
+
+impl ExecutorConfig {
+    /// Creates a new `ExecutorConfig` from a `timeout`. Note that the `timeout`
+    /// is *not* a global timeout, but will be applied to each individual call
+    /// of the underlying future.
+    /// The rest of the parameters - threshold, closed_len, etc, are initialized
+    /// to defaults.
+    pub fn new(timeout: Duration) -> Self {
+        let builder = Recloser::custom()
+            .error_rate(0.5)
+            .closed_len(100)
+            .half_open_len(10)
+            .open_wait(Duration::from_secs(5));
+        ExecutorConfig { builder, timeout }
+    }
+
+    /// The timeout for the individual executions of the future
+    pub fn timeout(mut self, timeout: Duration) -> Self {
+        self.timeout = timeout;
+        self
+    }
+
+    /// The threshold for opening the circuit, expressed as a float between 0 and 1.0.
+    /// The state of the CircuitBreaker changes from CLOSED to OPEN when the failure rate
+    /// is equal or greater than `threshold`.
+    /// For example when more than 50% (`0.5f32`) of the recorded calls have failed.
+    pub fn threshold(mut self, threshold: f32) -> Self {
+        self.builder = self.builder.error_rate(threshold);
+        self
+    }
+
+    /// How many calls will the circuit remain closed for before the failure
+    /// rate is recalculated and the circuit may open
+    pub fn closed_len(mut self, closed_len: usize) -> Self {
+        assert_ne!(closed_len, 0, "closed_len must be non-zero");
```

Comment:
1) seems like we should probably just std::num::NonZeroUsize instead 
2) shouldn't we return a Result with a new `ExecutorConfigError` type instead of asserting?